feat(replication): Add support for `TEMPORARY` replication slots

akselbor opened the pull request:
## What kind of change does this PR introduce?

~~This adds a new attribute `replication_slot: ReplicationSlotConfig` to `PipelineConfig` allowing users to choose whether to use regular (`ReplicationSlotConfig::Permanent`) or temporary (`ReplicationSlotConfig::Temporary`) replication slots. This is passed down to calls to `PgReplicationClient` where relevant (when creating/fetching slots).~~

This adds a new attribute `replication_slot: ReplicationSlotConfig` to `PipelineConfig` allowing users to specify additional  configuration that will be for the replication slot. Currently, we only have `ReplicationSlotConfig { temporary: bool }`. However, this pattern allows us to more easily extend with additional options in the future. Postgres provides several knobs to turn for replication slots ([docs](https://www.postgresql.org/docs/current/protocol-replication.html)):
* `LOGICAL` / `PHYSICAL` (probably irrelevant)
* `TEMPORARY` (implemented here)
* `FAILOVER`
* ...

By making this a `struct`, we can easily extend it with additional options down the line, with no/minimal breakage.

## What is the current behavior?

Only regular (non-temporary) replication slots are supported.

## What is the new behavior?

Users can choose whether to use persisted/permanent or temporary replication slots. Having access to temporary replication slots can be useful for ephemeral use cases (e.g., using CDC to maintain an in-memory cache of select tables for an otherwise stateless application).

## Additional context

* This is a (small) breaking change since it alters the `PipelineConfig`
* `cargo test --workspace --all-features --no-fail-fast --exclude etl-destination` works fine locally, but haven't tested the `etl-destinations` tests
~* I opted to use a simple boolean rather than the `ReplicationSlotConfig` for the `FullApiPipelineConfig` to avoid having to pull in `utoipa` as a dependency for the `etl-config` crate.~
* Happy to implement any changes requested


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **New Features**
  * Added replication slot persistence configuration, allowing users to specify whether slots should be temporary or permanent.
  * Enhanced pipeline configuration to include replication slot settings.

* **Improvements**
  * Improved Docker initialization script compatibility with both Docker Compose v2 and legacy versions.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

akselbor committed:
Add support for TEMPORARY replication slots

akselbor committed:
Minor comment fixes

akselbor committed:
Fix PipelineConfig for benches/table_copies.rs

akselbor committed:
Update .snaps

iambriccardo commented on etl-api/src/configs/pipeline.rs:
```diff
@@ -31,6 +31,7 @@ pub struct FullApiPipelineConfig {
     #[schema(example = "my_publication")]
     #[serde(deserialize_with = "crate::utils::trim_string")]
     pub publication_name: String,
+    pub temporary_replication_slot: bool,
```
Are there other kind of replication slots? Asking since having a boolean might give us less flexibility in the future to support other replication slots types.

akselbor replied:
Hmmm as per the Postgres docs ([here](https://www.postgresql.org/docs/current/protocol-replication.html)) there are some other knobs that can be turned when creating the replication slots:
```
CREATE_REPLICATION_SLOT slot_name [ TEMPORARY ] { PHYSICAL | LOGICAL output_plugin } [ ( option [, ...] ) ] 
Create a physical or logical replication slot. See [Section 26.2.6](https://www.postgresql.org/docs/current/warm-standby.html#STREAMING-REPLICATION-SLOTS) for more about replication slots.

slot_name
The name of the slot to create. Must be a valid replication slot name (see [Section 26.2.6.1](https://www.postgresql.org/docs/current/warm-standby.html#STREAMING-REPLICATION-SLOTS-MANIPULATION)).

output_plugin
The name of the output plugin used for logical decoding (see [Section 47.6](https://www.postgresql.org/docs/current/logicaldecoding-output-plugin.html)).

TEMPORARY
Specify that this replication slot is a temporary one. Temporary slots are not saved to disk and are automatically dropped on error or when the session has finished.

The following options are supported:

TWO_PHASE [ boolean ]
If true, this logical replication slot supports decoding of two-phase commit. With this option, commands related to two-phase commit such as PREPARE TRANSACTION, COMMIT PREPARED and ROLLBACK PREPARED are decoded and transmitted. The transaction will be decoded and transmitted at PREPARE TRANSACTION time. The default is false.

RESERVE_WAL [ boolean ]
If true, this physical replication slot reserves WAL immediately. Otherwise, WAL is only reserved upon connection from a streaming replication client. The default is false.

SNAPSHOT { 'export' | 'use' | 'nothing' }
Decides what to do with the snapshot created during logical slot initialization. 'export', which is the default, will export the snapshot for use in other sessions. This option can't be used inside a transaction. 'use' will use the snapshot for the current transaction executing the command. This option must be used in a transaction, and CREATE_REPLICATION_SLOT must be the first command run in that transaction. Finally, 'nothing' will just use the snapshot for logical decoding as normal but won't do anything else with it.

FAILOVER [ boolean ]
If true, the slot is enabled to be synced to the standbys so that logical replication can be resumed after failover. The default is false.
```

I would guess that `etl` will limit itself to logical replication, so the physical replication is irrelevant. Some of the others might be relevant, such as `FAILOVER` (to enable clean failover if someone is running a high-availability postgres cluster using e.g., `cloudnative-pg`). But iirc the `etl` currently couldn't support this without adding some kind of reconnect logic anyhow. 

But I agree with your point! It could make sense to make it more future-proof wrt. other existing (or future) replication slot options later, by changing the `ReplicationSlotConfig` to something like:
```
pub struct ReplicationSlotConfig {
    // Making this private has the benefit of forcing creation through the Default trait. If we combine it with some
    // builder methods we prevent future additions from being a breaking change. But could also just be pub
    temporary: bool,
    // extend with other attributes as needed in the future, e.g.,
    // failover: bool
}

impl Default for ReplicationSlotConfig { /* ... */ }

impl ReplicationSlotConfig { /* ... */ }
```

iambriccardo commented on etl-api/src/configs/pipeline.rs:
```diff
@@ -89,6 +94,7 @@ pub struct PartialApiPipelineConfig {
 #[derive(Debug, Clone, Serialize, Deserialize)]
 pub struct StoredPipelineConfig {
     pub publication_name: String,
+    pub replication_slot: ReplicationSlotConfig,
```
This will crash for existing stored configs.
```suggestion
    #[serde(default)]
    pub replication_slot: ReplicationSlotConfig,
```

This has to be paired with the other comment that I left.

iambriccardo reviewed (commented):
Thanks for the contribution! Are there other types of slots?

Left some comments around backward compatibility.

akselbor committed:
Change ReplicationSlotConfig from enum to struct

akselbor committed:
Update .snaps

akselbor committed:
Run cargo fmt

akselbor committed:
Change ReplicationSlotConfig to derive Default

akselbor committed:
cargo fmt

akselbor commented:
Changed from an enum to a structure, i.e.,
```
pub struct ReplicationSlotConfig {
    pub temporary: bool
}
```
with a default of non-temporary (`temporary: false`).

This should be better long-term in the event that we want to add additional options.

Also implemented the changes mentioned above.

akselbor committed:
Run clippy

iambriccardo commented:
Thanks for the updates, I will get to this PR once I finalize some other pieces of work!

iambriccardo commented on etl-config/src/shared/pipeline.rs:
```diff
@@ -90,3 +94,9 @@ impl From<PipelineConfig> for PipelineConfigWithoutSecrets {
+#[derive(Clone, Copy, Debug, Deserialize, Serialize, Default)]
+#[serde(rename_all = "snake_case")]
+pub struct ReplicationSlotConfig {
+    pub temporary: bool,
```
I have tried searching online for variants, to keep our format open for the future we could do something like:

```suggestion
    pub persistence: ReplicationSlotPersistence,
```

And then the `ReplicationSlotPersistence` could be something like:
```
enum ReplicationSlotType {
    Permanent,
    Temporary
}
```

And that could be serialized as an internally tagged enum as snake case and with the identifier named something like `type`.

The reason why I am suggesting this is that if there is ever the case where persistence changes or we want a slight variation, having a boolean is painful to migrate from. Unfortunately me being overly strict is because we store this config in the middleware db and the migration paths are very tricky.

iambriccardo reviewed (commented):
Left some other comments, just spent more time reviewing it closely.

Unfortunately given our current setup, we need to be very cautious about backward and forward compatibility.

akselbor commented:
Yeah I get that, no worries! I'll get to these when I have time (probably in a couple of days). Might throw in some tests for backwards compatibility as well (just need to do some reading into the API, given that I have only been consuming `etl` as a Rust library so far)

akselbor committed:
Merge in upstream and change to ReplicationSlotPersistence

akselbor committed:
Cargo fmt

akselbor committed:
Remove left-over schema example for replication slot config

akselbor committed:
Docs

akselbor committed:
Ignore PipelineBuilder doctest

akselbor committed:
Merge remote-tracking branch 'upstream/main'

akselbor committed:
Merge in upstream/main
